src/main/java/org/ow2/proactive/catalog/graphql/handler/catalogobject/CatalogObjectAndOrGroupFilterHandler.java

```diff
@@ -0,0 +1,237 @@
+                                                              .collect(Collectors.toList());
+
+            log.debug(collect);
+
+            Specification<CatalogObjectRevisionEntity> finalSpecification = buildFinalSpecification(collect);
+            return Optional.of(finalSpecification);
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * build up the final specification from the post-order traverse node result
+     * 
+     * @param collect
+     * @return
+     */
+    private Specification<CatalogObjectRevisionEntity>
+            buildFinalSpecification(List<CatalogObjectWhereArgsTreeNode> collect) {
+        Deque<Specification<CatalogObjectRevisionEntity>> stack = new LinkedList<>();
+
+        // node
+        for (CatalogObjectWhereArgsTreeNode argsTreeNode : collect) {
+
+            Operations nodeOperations = argsTreeNode.getOperations();
+
+            List<Specification<CatalogObjectRevisionEntity>> nodeSpecList = new ArrayList<>();
+
+            for (CatalogObjectWhereArgs whereArg : argsTreeNode.getWhereArgs()) {
+                if (whereArg.getOrArg() == null && whereArg.getAndArg() == null) {
+                    for (FilterHandler<CatalogObjectWhereArgs, CatalogObjectRevisionEntity> fieldFilterHandler : fieldFilterHandlers) {
+                        Optional<Specification<CatalogObjectRevisionEntity>> sp = fieldFilterHandler.handle(whereArg);
+                        if (sp.isPresent()) {
+                            nodeSpecList.add(sp.get());
+                            break;
+                        }
+                    }
+                } else {
+                    nodeSpecList.add(stack.pop());
+                }
+            }
+
+            if (nodeSpecList.isEmpty()) {
+                throw new IllegalArgumentException("At least one argument is needed");
+            }
+
+            Specification<CatalogObjectRevisionEntity> temp;
+            if (nodeOperations == Operations.AND) {
+                temp = AndSpecification.builder().fieldSpecifications(nodeSpecList).build();
+            } else {
+                temp = OrSpecification.builder().fieldSpecifications(nodeSpecList).build();
+            }
+
+            stack.push(temp);
+
+        }
+        return stack.pop();
+    }
+
+    /**
+     * /!\ NOTE : do not change this method without good reason and good tests
+     *
+     * @param andOrArgs
+     * @param operations
+     * @return
+     */
+    private List<CatalogObjectWhereArgsTreeNode>
+            postOrderTraverseWhereArgsToHaveTreeNodes(List<CatalogObjectWhereArgs> andOrArgs, Operations operations) {
+
+        // post-order traverse tree algorithm
+        Deque<List<CatalogObjectWhereArgs>> stack = new LinkedList<>();
+        stack.push(andOrArgs);
+
+        List<CatalogObjectWhereArgsTreeNode> ret = new ArrayList<>();
+
+        while (!stack.isEmpty()) {
+            List<CatalogObjectWhereArgs> top = stack.pop();
+            if (top != null) {
```

Comment:
Here, I prefer not to judge the top, but to judge top.left and top.right, which means that when top.left != null, we push top.left to the stack, then when top.right != null, we push top.right to the stack. To do so, we do not need to push so many null to the stack.